replications/service.go

```diff
@@ -2,53 +2,318 @@ package replications
+	"github.com/influxdata/influxdb/v2/replications/internal"
+	"github.com/influxdata/influxdb/v2/snowflake"
+	"github.com/influxdata/influxdb/v2/sqlite"
+	"github.com/mattn/go-sqlite3"
 )
 
-var errNotImplemented = &ierrors.Error{
-	Code: ierrors.ENotImplemented,
-	Msg:  "replication APIs not yet implemented",
+var (
+	errNotImplemented = &ierrors.Error{
+		Code: ierrors.ENotImplemented,
+		Msg:  "replication APIs not yet implemented",
+	}
+	errReplicationNotFound = &ierrors.Error{
+		Code: ierrors.ENotFound,
+		Msg:  "replication not found",
+	}
+)
+
+func errRemoteNotFound(id platform.ID, cause error) error {
+	return &ierrors.Error{
+		Code: ierrors.EInvalid,
+		Msg:  fmt.Sprintf("remote %q not found", id),
+		Err:  cause,
+	}
 }
 
-func NewService() *service {
-	return &service{}
+func errLocalBucketNotFound(id platform.ID, cause error) error {
+	return &ierrors.Error{
+		Code: ierrors.EInvalid,
+		Msg:  fmt.Sprintf("local bucket %q not found", id),
+		Err:  cause,
+	}
 }
 
-type service struct{}
+func NewService(store *sqlite.SqlStore, bktSvc influxdb.BucketService) *service {
+	return &service{
+		store:         store,
+		idGenerator:   snowflake.NewIDGenerator(),
+		bucketService: bktSvc,
+		validator:     internal.NewValidator(),
+	}
+}
+
+type ReplicationValidator interface {
+	ValidateReplication(context.Context, *internal.ReplicationHTTPConfig) error
+}
+
+type service struct {
+	store         *sqlite.SqlStore
+	idGenerator   platform.IDGenerator
+	bucketService influxdb.BucketService
+	validator     ReplicationValidator
+}
 
 var _ influxdb.ReplicationService = (*service)(nil)
 
 func (s service) ListReplications(ctx context.Context, filter influxdb.ReplicationListFilter) (*influxdb.Replications, error) {
-	return nil, errNotImplemented
+	q := sq.Select(
+		"id", "org_id", "name", "description",
+		"remote_id", "local_bucket_id", "remote_bucket_id",
+		"max_queue_size_bytes", "current_queue_size_bytes",
+		"latest_response_code", "latest_error_message").
+		From("replications").
+		Where(sq.Eq{"org_id": filter.OrgID})
+
+	if filter.Name != nil {
+		q = q.Where(sq.Eq{"name": *filter.Name})
+	}
+	if filter.RemoteID != nil {
+		q = q.Where(sq.Eq{"remote_id": *filter.RemoteID})
+	}
+	if filter.LocalBucketID != nil {
+		q = q.Where(sq.Eq{"local_bucket_id": *filter.LocalBucketID})
+	}
+
+	query, args, err := q.ToSql()
+	if err != nil {
+		return nil, err
+	}
+
+	var rs influxdb.Replications
+	if err := s.store.DB.SelectContext(ctx, &rs.Replications, query, args...); err != nil {
+		return nil, err
+	}
+	return &rs, nil
 }
 
 func (s service) CreateReplication(ctx context.Context, request influxdb.CreateReplicationRequest) (*influxdb.Replication, error) {
-	return nil, errNotImplemented
+	s.store.Mu.Lock()
+	defer s.store.Mu.Unlock()
+
+	// Ensure the referenced bucket exist locally.
+	// NOTE: Since bucket metadata is stored in bolt, we don't have a FK in the DB to guard against
```

Comment:
What is bolt? I'm not familiar with this.  Does bolt not do foreign keys?

Reply:
Bolt is an embedded KV store, it doesn't have foreign keys like a relational DB would. The cross-referencing here is complicated further because the new metadata about replications is stored in sqlite, not bolt.

Reply:
Thank you for taking the time to break this down for a newbie.  I did the quick google for Bolt KV Store and the first result was for a git repo that is archived.  Is this project using that archived version of Bolt or something else? My bosses and mentors keep telling me to avoid unsupported or end of life software.

Reply:
We use a fork of the original maintained by the etcd organization, see https://github.com/etcd-io/bbolt

Reply:
Thanks again for explaining more about Bolt! Was there somewhere in the readme that I should have read that explains the usage of the forked Bolt repository?